### PR TITLE
gh-98415: Fix UUID ifconfig MAC address detection

### PR DESCRIPTION
**Pull request by csanders-git**
See #98415 for explanation

* Issue: gh-98415

### CODE REVIEW

File: Lib/uuid.py [hunk 1]
```diff
@@ -371,7 +371,13 @@ def _get_command_stdout(command, *args):
...
         env = dict(os.environ)
         env['LC_ALL'] = 'C'
-        proc = subprocess.Popen((executable,) + args,
+        # Empty strings will be quoted by popen so we should just ommit it
+        command = (executable,)
+        if args == ('',):
+            command = (executable,)
+        else:
+            command = (executable,) + args
+        proc = subprocess.Popen(command,
```
Comment (corona10) [inline Lib/uuid.py]:
```suggestion
        # Empty strings will be quoted by popen so we should just ommit it
        command = (executable,)
        if args != ('',):
            command = (executable,) + args
        proc = subprocess.Popen(command,
```

**Review by corona10 — CHANGES_REQUESTED** [review]
Nice catch!

I reproduce the issue with the following script:


````python

import subprocess
"""Get the hardware address on Unix by running ifconfig."""
# This works on Linux ('' or '-a'), Tru64 ('-av'), but not all Unixes.

def foo(*args):
    command = ("/sbin/ifconfig",) + args
    proc = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.DEVNULL)
    if not proc:
        return
    stdout, stderr = proc.communicate()
    mac = stdout # This is fine for this example (normally it'd have parsed out a mac
    if not mac:
        print(args, mac, stderr)


for args in ('', '-a', '-av'):
    foo(args)

````

File: Misc/NEWS.d/next/Core and Builtins/2022-10-19-01-01-08.gh-issue-98415.ZS2eWh.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Fix detection of MAC addresses for UUID on certain OSs . Patch by Chaim Sanders
```
Comment (corona10) [inline Misc/NEWS.d/next/Core and Builtins/2022-10-19-01-01-08.gh-issue-98415.ZS2eWh.rst]:
```suggestion
Fix detection of MAC addresses for :mod:`uuid` on certain OSs. Patch by Chaim Sanders
```

File: Lib/uuid.py [hunk 3]
```diff
@@ -371,7 +371,11 @@ def _get_command_stdout(command, *args):
         # for are actually localized, but in theory some system could do so.)
         env = dict(os.environ)
         env['LC_ALL'] = 'C'
-        proc = subprocess.Popen((executable,) + args,
+        # Empty strings will be quoted by popen so we should just ommit it
+        command = (executable,)
+        if args != ('',):
+            command = (executable,) + args
```
Comment (corona10) [inline Lib/uuid.py]:
```suggestion
        # Empty strings will be quoted by popen so we should just ommit it
        if args != ('',):
            command = (executable,) + args
        else:
            command = (executable,)
```

**Comment by vstinner** [comment]
I have a preference for ``(executable, *args)`` syntax when building command line arguments :-) IMO it's more readable.

**Review by vstinner — COMMENTED** [review]
Would it be possible to write an unit test with unittest.mock to check that subprocess is not called with an empty string argument?

File: Lib/uuid.py [hunk 4]
```diff
@@ -371,7 +371,12 @@ def _get_command_stdout(command, *args):
         # for are actually localized, but in theory some system could do so.)
         env = dict(os.environ)
         env['LC_ALL'] = 'C'
-        proc = subprocess.Popen((executable,) + args,
+        # Empty strings will be quoted by popen so we should just ommit it
+        if args != ('',):
+            command = (executable,) + args
```
Comment (corona10) [inline Lib/uuid.py]:
```suggestion
            command = (executable, *args)
```

**Comment by corona10** [comment]
> Would it be possible to write an unit test with unittest.mock to check that subprocess is not called with an empty string argument?

It looks hard to write the unittest due to system-specific implementation.
I would like to skip the writing test code at this moment,

**Comment by vstinner** [comment]
Ok, thanks @csanders-git for the fix: I merged it. I rewrote the commit message to clarify its scope.

**Comment by vstinner** [comment]
@corona10:
> It looks hard to write the unittest due to system-specific implementation.
> I would like to skip the writing test code at this moment,

It makes me sad, but I'm fine with not adding a test :-)

**Comment by vstinner** [comment]
In the past, I fixed a risk of shell injection by avoiding a shell in the uuid module: commit b9d0199c073fc963b27106f4b7752dcbe39bfed3. I didn't add any test for that neither :-(

**Pull request merged**